contracts/generic-composable-tasks/src/call_dispatcher.rs

```diff
@@ -0,0 +1,181 @@
+                gas_limit,
+                function_name: exec_arg.function_name,
+                args: exec_arg.args,
+            });
+
+            match payment_type {
+                PaymentType::None => {
+                    let opt_back_transfers =
+                        self.perform_no_transfer(dest_address, call_type, opt_raw_call_args);
+                    opt_last_back_transfers = self.handle_back_transfers_if_any(
+                        &mut total_egld,
+                        &mut all_esdt,
+                        opt_back_transfers,
+                    );
+                }
+                PaymentType::Egld { amount } => {
+                    let opt_back_transfers = self.perform_egld_transfer(
+                        dest_address,
+                        call_type,
+                        amount,
+                        opt_raw_call_args,
+                    );
+                    opt_last_back_transfers = self.handle_back_transfers_if_any(
+                        &mut total_egld,
+                        &mut all_esdt,
+                        opt_back_transfers,
+                    );
+                }
+                PaymentType::FixedPayments { esdt_payments } => {
+                    let opt_back_transfers = self.perform_multi_esdt_transfer(
+                        dest_address,
+                        call_type,
+                        esdt_payments,
+                        opt_raw_call_args,
+                    );
+                    opt_last_back_transfers = self.handle_back_transfers_if_any(
+                        &mut total_egld,
+                        &mut all_esdt,
+                        opt_back_transfers,
+                    );
+                }
+                PaymentType::ReceivedPaymentsFromSc => {
+                    require!(
+                        opt_last_back_transfers.is_some(),
+                        "No payments received from SC"
+                    );
+
+                    // only works with ESDTs
+                    let last_back_transfers = unsafe { opt_last_back_transfers.unwrap_unchecked() };
+                    for transfer in &last_back_transfers.esdt_payments {
+                        let deduct_result = all_esdt.deduct_payment(&transfer);
+                        require!(deduct_result.is_ok(), "May not use these back transfers");
+                    }
+
+                    let opt_back_transfers = self.perform_multi_esdt_transfer(
+                        dest_address,
+                        call_type,
+                        last_back_transfers.esdt_payments,
+                        opt_raw_call_args,
+                    );
+                    opt_last_back_transfers = self.handle_back_transfers_if_any(
+                        &mut total_egld,
+                        &mut all_esdt,
+                        opt_back_transfers,
+                    );
+                }
+            }
```

Comment:
this can be extracted into a function `handle_call` or something similar having `arg`, `&mut total_egld`, `&mut all_esdt` and `&mut opt_last_back_transfers` as params

It makes the for loop look cleaner